neuralpp/test/quick_tests/symbolic/normalizer_test.py

```diff
@@ -78,3 +79,50 @@ def test_normalizer3():
     assert isinstance(context, Z3SolverExpression)
     result = normalizer.normalize(expr, context)
     assert result.structure_eq(if_then_else(c, 45, if_then_else(a > b, f(False, 66), 3)))
+
+
+def test_quantifier_normalizer():
+    from neuralpp.symbolic.constants import int_add, int_multiply
+    i = BasicVariable('i', int)
+    empty_context = Z3SolverExpression()
+    i_range = empty_context & (0 < i) & (i < 10)
+    sum_ = BasicSummation(int, i, i_range, i)
+
+    context = empty_context
+    normalizer = Normalizer()
+    assert normalizer.normalize(sum_, context).structure_eq(sum_)
+
+    context = empty_context & (i < 5)
+    with pytest.raises(ValueError):
+        normalizer.normalize(sum_, context)
+
+    j = BasicVariable('j', int)
+    sum_ = BasicSummation(int, i, empty_context & (j < i) & (i < 10), i + j)
+    context = empty_context & (j == 5)
+    assert normalizer.normalize(sum_, context).structure_eq(
+        BasicSummation(int, i, empty_context & (j < i) & (i < 10), 5 + i))  # SymPy switched 5 and i
```

Comment:
🤔 Actually I was just thinking if it would make sense to replace `j` in the constrain as well. Do you have any thoughts on this?

Reply:
I think both are correct. If we want to have the "replacing according to the context" behavior, what we essentially need is "simplify constrain according to context", which is more general and not only deals with equalities. 

But since `constrain` is a context (`Z3SolverExpression`), simplification of `constrain` is not very efficient (first convert to `SymPyExpression`, then we need to manually convert it back to `Z3SolverExpression`, which is not implemented yet). 

I would argue since this feature (i) does not affect correctness, (ii) requires some non-trivial work to implement, and (iii) would not be very efficient, we should just be lazy here and stick to the current behavior.